[16.0] shopfloor: location_content_transfer: get_work additional domain and sort key.

lmignon opened the pull request:
The initial aim of this PR is to enable the 'location_content_transfer' scenario to be configured more precisely. 
It aims to:

* allow additional criteria to be configured to be applied when the user searches for work to be done (example of use: a menu for processing urgent transfers only)
* allow you to configure a different sort to be applied to the operations to be performed than the one based on priority. (by location, for example)
 
When analysing the code, I found that there was an inconsistency between the code used to create the menu counters and the one used to find the transactions in the 'location_content_transfer' scenario. The code in the latter scenario was functionally identical to the one used for menus and zone_picking. Harmonisation work has therefore been carried out to eliminate this inconsistency.

As result, a generic solution has been put in place for scenarios using the 'search_move_line' action to find the operations to be carried out, so that the menus can be used to configure an additional domain to be applied by default and the method for sorting operations.

lmignon commented:
cc @jbaudoux @rousseldenis

lmignon commented:
This way of doing things is not the most performant one. It requires to browse all the candidate pickings and related stock.move.line for ordering the operations and return the location of the first one into the list. The proposed implementation should not alter the current performance since the initial search was previously done without 'limit' criteria and all the related `stock.move.line` were loaded by the system due to the orm's pre-fetching mechanism. Another approach might have been not to allow a sorting method to be entered in the menu config and to adapt the code to sort via the orm. In our case, we wanted to order based on the source location of the operation. This implies a join between the stock_picking and stock_move_line tables for sorting. Such a join is only possible by manipulating the orm's low-level classes. This approach makes the code more complex, but also offers the advantage of using these low-level classes to obtain the information required in 1 SQL query. If you think it's better to go this way, here's what some of the code might look like https://github.com/acsone/wms/commit/04d37a0ac19f1e06d23285eed50d904fadb4babf

lmignon committed:
[FIX] shopfloor: menu counters

The priority lines count must be the sum of operations linked to a priority move, while the priority pickings count must be the sum of priority pickings.

lmignon committed:
[FIX] shopfloor: search_move_line._sort_key_move_lines

Ensures opertations assigned to current user are always first when sorting lines

lmignon committed:
[FIX] shopfloor: search_move_line._search_move_lines_by_location_domain

Never return operations from pickings assigned to an other user when the method is called with the parameter 'match_user' set to True

lmignon committed:
[IMP] shopfloor: search_move_line: Makes location optional

If a location is not provided we take the one from the picking type by default

lmignon committed:
[IMP] shopfloor: location_content_transfer: Uses search_move_line

The scenario now uses the 'search_move_line' action to search for operations to perform when the user requests work. This eliminates duplicate code between the scenario and the 'search_move_line' action.

lmignon committed:
[IMP] shopfloor: Custom sort and additional domain

For scenario using the 'search_move_line' action to get lines
to work on, it's now possible to provide an additional domain
and a custom sort key code when configuring menu using these scenario.
This applies to the location_content_transfer scenario and partialy to
the zone_picking one

rousseldenis committed:
[FIX] shopfloor: Restore hook to select first move line for content transfer

jbaudoux commented on shopfloor/actions/move_line_search.py:
```diff
@@ -31,49 +67,58 @@ def _search_move_lines_by_location_domain(
         picking_ready=True,
         # When True, adds the package in the domain even if the package is False
         enforce_empty_package=False,
+        enforce_picking_types=True,
```
You are not enforcing the picking types in the domain when its empty. That's not inline with `enforce_empty_package`.
I would drop this new argument

lmignon replied:
@jbaudoux This new argument is required to allow to check if some operations are already planned into a location when you ask for work on a location with the location_content_transfer scenario configured. If you filter on the picking type at this stage, you will not get the message 'You cannot move this using this menu.'  but you will receive 'New move lines cannot be assigned: canceled'

jbaudoux replied:
But the naming of the argument is confusing because it looks like the other existing enforce argument but reacts differently

lmignon replied:
I am open to any suggestion to change the name of this parameter :smirk:

jbaudoux replied:
I still don't understand why we need such argument

jbaudoux commented on shopfloor/actions/move_line_search.py:
```diff
@@ -31,49 +67,58 @@ def _search_move_lines_by_location_domain(
+            picking_type or self.picking_types
+            if enforce_picking_types
+            else picking_type
+        )
```
```suggestion
        picking_types = picking_type or self.picking_types
```

jbaudoux replied:
I still think this is enough

jbaudoux commented on shopfloor/actions/move_line_search.py:
```diff
@@ -82,38 +127,80 @@ def search_move_lines_by_location(
-    def _sort_key_move_lines(order):
+    def _sort_key_move_lines(self, order=None):
         """Return a sorting function to order lines."""
+        if order == "none":
```
?
```suggestion
        if order is None:
```

jbaudoux commented on shopfloor/actions/move_line_search.py:
```diff
@@ -82,38 +127,80 @@ def search_move_lines_by_location(
+        if user_id == self.env.uid:
+            priority = 0
+        elif user_id is None:
+            priority = 1
+        else:
+            priority = 2
```
```suggestion
            priority = -1
        else:
            priority = -2
```

lmignon replied:
We want those assigned to current user first then the one unassigned ...
![image](https://github.com/user-attachments/assets/cbb196e5-74f5-41ab-a929-dec1fe7a0029)

jbaudoux commented on shopfloor/services/location_content_transfer.py:
```diff
@@ -252,24 +235,16 @@ def _create_moves_from_location(self, location):
+        for move_line in move_lines:
+            if move_line.location_id:
+                return move_line.location_id
         return location
```
`location_id` is a required field. You can just do
```suggestion
        return fields.first(move_lines).location_id
```

jbaudoux commented on shopfloor/static/description/index.html:
```diff
@@ -8,10 +8,11 @@
 /*
 :Author: David Goodger (goodger@python.org)
-:Id: $Id: html4css1.css 8954 2022-01-20 10:10:25Z milde $
+:Id: $Id: html4css1.css 9511 2024-01-13 09:50:07Z milde $
```
Can you skip the generation of auto generated files?

jbaudoux reviewed (changes requested):
Some remarks

rousseldenis commented:
@jbaudoux

jbaudoux commented on shopfloor/services/zone_picking.py:
```diff
@@ -429,7 +429,7 @@ def _find_location_move_lines(
-        return self.search_move_line.search_move_lines_by_location(
+        return self.search_move_line.search_move_lines(
             locations or self.zone_location,
             picking_type=picking_type or self.picking_type,
```
Already managed by `search_move_lines`
```suggestion
            picking_type=picking_type,
```

jbaudoux replied:
what about this one ?

jbaudoux commented on shopfloor/actions/move_line_search.py:
```diff
@@ -31,49 +67,58 @@ def _search_move_lines_by_location_domain(
+            else picking_type
+        )
+        if picking_types:
             domain += [("picking_id.picking_type_id", "in", self.picking_types.ids)]
```
```suggestion
            domain += [("picking_id.picking_type_id", "in", picking_types.ids)]
```

jbaudoux replied:
I think you forgot that one

jbaudoux commented on shopfloor/services/location_content_transfer.py:
```diff
@@ -356,7 +328,10 @@ def scan_location(self, barcode):  # noqa: C901
 
-        move_lines = self._find_location_move_lines_from_scan_location(location)
+        move_lines = self.search_move_line.search_move_lines(
+            locations=location, match_user=True, enforce_picking_types=False
```
We should still respect the picking types of the menu
```suggestion
            locations=location, match_user=True,
```

jbaudoux reviewed (changes requested):
I still don't get the purpose of `enforce_picking_types` :/

lmignon commented:
@jbaudoux If I remove the possibility to search for picking in an other picking type than the one provided by the menu if specified, I'm no more able to support the use case in the location_content_transfer scenario when content has different picking type than menu, and the menu is configured with the option to unreserve. In such a case, content must be unreserved, new moves created and the previous content re-reserved. That why I need this new `enforce_picking_types` parameter to also uses the search method in such a case where I don't want to enforce the picking types to the those configured on the menu or passed to the method.

lmignon committed:
[IMP] shopfloor: Optionnal picking type on moves search

To disable the filtering of move line based on specific picking types, the additional parameter ‘enforce_picking_types’ had to be set to True when calling the method 'search_move_lines' of the 'hopfloor.search.move.line' action. The name of this parameter was not really relevant and its meaning conflicted with another parameter ‘enforce_package_types’. To avoid this situation, the semantics of the ‘picking_type’ parameter have been modified so that it can be explicitly stated that the search is not on a particular picking type if an empty recordset is passed as a parameter. Previously, if the parameter was None or an empty recordset, the filtering was forced on the picking types specified into the work context.

Translated with DeepL.com (free version)

lmignon commented:
@jbaudoux Your last comment is now implemented into the last commit.

jbaudoux commented:
> @jbaudoux Your last comment is now implemented into the last commit.

@lmignon nearly. There are still 2 unresolved comments related to your last changes. Should be obvious

lmignon commented:
@jbaudoux done

jbaudoux commented on shopfloor/__manifest__.py:
```diff
@@ -6,11 +6,11 @@
     "category": "Inventory",
     "website": "https://github.com/OCA/wms",
-    "author": "Camptocamp, BCIM, Akretion, Odoo Community Association (OCA)",
+    "author": "Camptocamp, BCIM, Akretion, ACSONE SA/NV, Odoo Community Association (OCA)",
```
Can you drop this change from this PR?

lmignon replied:
I will remove the commit with this change. I'm a little bit surprised that with all the work we do on this addon we can't put ACSONE into the authors list.

lmignon replied:
commit removed

jbaudoux commented:
/ocabot merge nobump
